### PR TITLE
[asl] Global variables continued

### PR DESCRIPTION
**Pull request by maranget**
The last commit 7ed7265f08629022165a727218dfda9af0259e8a is a followup to PR #598: a traditional implementation of scope and heap.

All other commits are the ones of PR #598, up to some rebase operations of master.

### CODE REVIEW

**Comment by maranget** [comment]
> Hi @maranget, great that you have been able to finish your work! 🎉
> 
> I'm wondering about the best way to merge this with respect to #605: there will be a lot of conflict in `Interpreter.ml`, especially after [6fcfdf8](https://github.com/herd/herdtools7/commit/6fcfdf8695618695a179df2dc80c0b1cfc83f6d3).
> 
> I see various ways to do that:
> 
> 1. I try to rebase slowly my ~13 commits after [7ed7265](https://github.com/herd/herdtools7/commit/7ed7265f08629022165a727218dfda9af0259e8a)
> 2. I try to cherry-pick your commit after [Further work on ASL typing #605](https://github.com/herd/herdtools7/pull/605), maybe by breaking it in 2 between the changes to the variable semantics and the dependency updates (of course keeping you as Author)
> 
> What do you think?

Hi @HadrienRenaud, my initial idea was for you to cherry-pick 7ed7265f08629022165a727218dfda9af0259e8a on top of your PR #598.  Then (have a look at AS/unknown.litmus though), you can probably merge in master and attempt to rebase  #605 with argument master. It may be that writing code by hand to implement exception finally is the most simple way to proceed. This may not be as horrible as it seems...

**Comment by maranget** [comment]
To be clear, the last two commits of PR #598 are obsolete as their functionality is included in commit 7ed7265. You can probably get rid of them before cherry-picking 7ed7265.

**Comment by maranget** [comment]
Ah I see both #605 and #607 extends PR #598.  Basically, if I were you, I'd first try to insert 7ed7265 after ba367b8 in PR #605.

**Review by HadrienRenaud — COMMENTED** [review]
A few questions on this commit before I try and merge this.

---

> That is the runtime environment binds variables to "pointers", while the memory binds pointer to values.

I see that you have to free the explicit memory at each end of block. I wonder then what are the differences between an implementation with an environment that you free at each end of block (by also keeping the local `declared` variables) and this one? What is the purpose of having this extra step `name -> pointer -> value` compared to `name -> value`? (it is more standard is a nice reason why, but I'd like to understand if there are more fundamental reasons)

To be more explicit, if in env we had `type storage = C.v IMap.t`, then the `pop_local` implementation could be:
```ocaml
let pop_local old env =
  let global = env.global
  and local =
    let storage =
      let fold_declared storage x =
        match IMap.find_opt x old.local.storage with
        | None -> IMap.remove x storage
        | Some old_v -> IMap.add x old_v storage
       in
       List.fold_left fold_declared env.local.storage env.local.declared
     in
     { old.local with storage }
   in
   { global; local }
```

---

> One could also check that no side effect has been performed.

How would you check that the env has not been modified? With something like `Env.equal`? Or keeping an internal counter in the env that counts the number of changes, thus "the env not been modified" can be reduced to "this counter has not been incremented"? Or having two interpretation functions: one for side-effect-free expressions/statements, one for possible side-effects?

---

Those questions (and the comments in the code) makes me think that I need a bit of time to properly understand and rebase the code. I will thus prepare a with #605 and #607 merged for when you are back @maranget.

As a consequence, I think I can choose either strategy to merge (rebase #605 on #607 or the opposite). As time is not a constraint anymore, I think I will choose to cherry-pick this commit (probably in multiple commits) on top of #605.

Would that work for you @maranget? Any other thoughts?

**Comment by maranget** [comment]
> A few questions on this commit before I try and merge this.
> 
> > That is the runtime environment binds variables to "pointers", while the memory binds pointer to values.
> 
> I see that you have to free the explicit memory at each end of block. I wonder then what are the differences between an implementation with an environment that you free at each end of block (by also keeping the local `declared` variables) and this one? What is the purpose of having this extra step `name -> pointer -> value` compared to `name -> value`? (it is more standard is a nice reason why, but I'd like to understand if there are more fundamental reasons)
> 
> To be more explicit, if in env we had `type storage = C.v IMap.t`, then the `pop_local` implementation could be:
> 
> ```ocaml
> let pop_local old env =
>   let global = env.global
>   and local =
>     let storage =
>       let fold_declared storage x =
>         match IMap.find_opt x old.local.storage with
>         | None -> IMap.remove x storage
>         | Some old_v -> IMap.add x old_v storage
>        in
>        List.fold_left fold_declared env.local.storage env.local.declared
>      in
>      { old.local with storage }
>    in
>    { global; local }
> ```
> 

Well,removing the mem bindings of discarded variables (the ones declared  in block) is optional. I have added as a second thought to avoid having environments whose size grow at each loop iteration This is a sort of gc, while in your setting it would be mandatory.

Also observe that if it is possible to declare a new binding for a variable that is already in scope, the one map implementation is slightly more tricky:  if the discarded variable was bound then restore it to its value before block start, otherwise destroy the binding.

But overall you are right, a one-map implementation is possible. Feel free to choose this solution.

> > One could also check that no side effect has been performed.
> 
> How would you check that the env has not been modified? With something like `Env.equal`? Or keeping an internal counter in the env that counts the number of changes, thus "the env not been modified" can be reduced to "this counter has not been incremented"? Or having two interpretation functions: one for side-effect-free expressions/statements, one for possible side-effects?

I was thinking about recording all variable assigned, a bit like all variables declared already are. But with different rules for reset and transmission. A counter of assignments or even a boolean  look to be  sufficient abstractions.

However notice that identifying data races between the arguments of function and primitive calls would call for recording list of writes and reads of mutable  variables.

Alternatively a simple static analysis would provide an approximation that would be close to the runtime detection. A very simple analysis (no fixpoint, function call => side effect) may be sufficient in practice.


> 
> Those questions (and the comments in the code) makes me think that I need a bit of time to properly understand and rebase the code. I will thus prepare a with #605 and #607 merged for when you are back @maranget.
> 

I am back in two weeks.


> As a consequence, I think I can choose either strategy to merge (rebase #605 on #607 or the opposite). As time is not a constraint anymore, I think I will choose to cherry-pick this commit (probably in multiple commits) on top of #605.
> 
> Would that work for you @maranget? Any other thoughts?

My initial suggestion of inserting my commit after our common prefix may not be 100% rational. As you have written several commits after the fork I thought it was easier for you to replay them  on a modified interpreter, than operating  the other way round. I may not change much in practice.

File: asllib/AST.mli [hunk 1]
```diff
@@ -197,11 +197,15 @@ type local_decl_item =
...
+type assign_t =
+  | Assign
+  | Decl
+
 type stmt_desc =
   | S_Pass
   | S_Then of stmt * stmt
   | S_Decl of local_decl_keyword * local_decl_item * expr option
-  | S_Assign of lexpr * expr
+  | S_Assign of assign_t * lexpr * expr
```
[the remark this answers is not in the document]
Comment (maranget) [reply asllib/AST.mli:208]: Yes, we should. Nothing fundamental here.

File: asllib/Interpreter.ml [hunk 2]
```diff
@@ -87,12 +82,11 @@ module Make (B : Backend.S) (C : Config) = struct
...
   let prod_map f =
-    let one acc elt =
+    let one elt acc =
       let* v = f elt and* acc = acc in
-      return (v :: acc)
-    in
-    function
-    | [] -> return [] | li -> List.fold_left one (return []) li >=> List.rev
+      return (v :: acc) in
+    fun xs -> List.fold_right one xs (return [])
```
[the remark this answers is not in the document]
Comment (maranget) [reply asllib/Interpreter.ml:88]: One vs. two function call, the latter is tail-recursive the former not.

File: asllib/Interpreter.ml [hunk 3]
```diff
@@ -290,86 +285,158 @@ module Make (B : Backend.S) (C : Config) = struct
...
+            let* v1,_ = m1 and* v2,_ = m2 in
+            return ((v1,v2),env)))
+
+  let eval_par eval_expr env e1 e2 = do_par env eval_expr e1 eval_expr e2
+
+  let rec eval_par_list f env = function
+    | [] -> return ([],env)
+    | x::xs ->
+       let* (y,ys),env = do_par env f x (eval_par_list f) xs in
+       return ((y::ys),env)
```
[the remark this answers is not in the document]
Comment (maranget) [reply asllib/Interpreter.ml:377]: The resultig dependencies are different. `eval_par_list` introduce no dependencies between the list elements, while `eval_list_m` sequence them left to right with bind_seq.

Using `eval_par_list` everywhere is an option I would favor.

**Comment by HadrienRenaud** [comment]
Thank you for the explanation, they are much helpful.

> Well,removing the mem bindings of discarded variables (the ones declared in block) is optional. I have added this phase as a second thought, in order to avoid having environments whose size grow at each loop iteration This is a sort of gc. While in your setting, it would be mandatory.
> 
> Also observe that if it is possible to declare a new binding for a variable that is already in scope, the one map implementation is slightly more tricky: if the discarded variable was bound then restore it to its value before block start, otherwise destroy the binding. By contrast variables that have no been declared in the block but have been assigned have to keep their updated values.
> 
> But overall you are right, a one-map implementation is possible. Feel free to choose this solution.

I will try both and propose the two. We can choose when you come back, but as I think I understand a bit more what is happening, both solutions are totally ok from my point of view.

With your solution, we could even do the "gc collect" only when going out of a loop scope. This way we would have only minimal impact. I'm starting to think that this might be the simplest solution for the general case.

> Alternatively a simple static analysis would provide an approximation that would be close to the runtime detection. A very simple analysis (no fixpoint, function call => side effect) may be sufficient in practice.

I like this static analysis idea, because I think we should flag those error at type-checking time as much as possible.
However it might not be as simple, as the following code should in theory be legal:
```
func incr(x: integer) => integer begin return x + 1; end
func main() begin
  let x = 0[incr(3)..incr(0)];
end
```
Given this, some thinking will have to go into flagging some functions/primitives as side-effect-free, or even which side-effects they introduce. It might also help checking at type-check time weather two sub-expressions have conflicting side-effects.

I think this work is very interesting, but will come after the semantics already work.

**Comment by maranget** [comment]
Indeed, I agree on focusing first on semantics correctness and then on optimisations. By the way, I would support the idea of evaluating all arguments in parallel, that is to replace `eval_list_m` and `eval_bind_m` by adapted `eval_par_..` functions.

File: asllib/Interpreter.mli [hunk 4]
```diff
@@ -20,7 +20,8 @@
 module type S = sig
   module B : Backend.S
 
-  type body = B.value B.m list -> B.value list B.m
+  type 'a m = 'a B.m
```
[the remark this answers is not in the document]
Comment (maranget) [reply asllib/Interpreter.mli:23]: Ok

File: asllib/Interpreter.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (HadrienRenaud) [reply asllib/Interpreter.ml:377]: I like the idea of keeping `bind_seq` for when there is a real `;` between two things.

We have to be careful not to break any AArch64 `iico_order` though.

File: asllib/Interpreter.ml [hunk 5]
```diff
@@ -290,86 +285,158 @@ module Make (B : Backend.S) (C : Config) = struct
...
+  let rec eval_bind_list_m eval_expr env xes =
+    match xes with
+    | [] -> return ([],env)
+    | (x,e)::xes ->
+       B.delay
+         (eval_expr env e)
+         (fun (_,env) mv ->
+           let mv = mv >>= fun (v,_) -> return v in
+           let*| mvs,env = eval_bind_list_m eval_expr env xes in
+           return ((x,mv)::mvs,env))
```
[the remark this answers is not in the document]
Comment (HadrienRenaud) [reply asllib/Interpreter.ml:360]: Ok!

File: asllib/Interpreter.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (HadrienRenaud) [reply asllib/Interpreter.ml:88]: Okay, I will try and use this rationale

**Comment by HadrienRenaud** [comment]
## Update on this PR

### What has been done

I have taken the commit 7ed7265 and split it into 3 commits on #605 :
 - 5bac7042dc9f62d08c50ebe5b0562e23a8ed7617 that implements the correct dependencies, using `delay` as suggested.
 - 3d8a634f34d28e2f56a74e976b1e01219e338e38 that implements side-effects in conditionnal expressions.
 - de14d67fe35a3e90822873ba4cc0a540a57586e2 that implements the indirection in the memory storage.

Those commits have @maranget as author and me as co-author.

Furthermore, I have refactored `env.ml` and added @maranget explanations on the indirection in the memory storage in commit 28d2a3d8027dad92019fd124551a53998d983ca5.

### Next steps

IMHO

1. We should keep the PR or the branch open for a while to keep as a reference.
2. After a while, this PR can be closed, and maybe the branch deleted.
3. We should put it back to draft, as anyway it now requires rebasing on master.
4. #605 is now the candidate for merging on the ASL side.

**Pull request closed without merging**